### PR TITLE
OTT-1297: cache fix: populate cache only upon successful DB calls

### PR DESCRIPTION
**Pull request by pm-jaydeep-mohite**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

**Review by pm-viral-vala — APPROVED** [review]
Please create a new ticket for incorporating changes of using stale data from the cache and using the actual version value for cache keys.

File: modules/pubmatic/openwrap/cache/gocache/partner_config.go [hunk 1]
```diff
@@ -95,5 +96,8 @@ func (c *cache) getActivePartnerConfigAndPopulateWrapperMappings(pubID, profileI
 		c.metricEngine.RecordDBQueryFailure(queryType, strconv.Itoa(pubID), strconv.Itoa(profileID))
 		err = errorWrap(err, errAdunitConfig)
 	}
+	if errWrapperSlotMapping == nil && errAdunitConfig == nil {
+		c.cache.Set(cacheKey, partnerConfigMap, getSeconds(c.cfg.CacheDefaultExpiry-60))
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/partner_config.go:100]: Would be clean idea to add a new filed `c.cachetimeout` and set it once during `gocache.New()` call and use it here
```
		c.cache.Set(cacheKey, partnerConfigMap, getSeconds(c.cachetimeout))
```

File: modules/pubmatic/openwrap/database/mysql/adunit_config.go [hunk 2]
```diff
@@ -20,7 +22,7 @@ func (db *mySqlDB) GetAdunitConfig(profileID, displayVersion int) (*adunitconfig
 
 	var adunitConfigJSON string
 	err := db.conn.QueryRow(adunitConfigQuery).Scan(&adunitConfigJSON)
-	if err != nil {
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/database/mysql/adunit_config.go:25]: Add a UT to cover this change

**Review by pm-nilesh-chate — COMMENTED** [review]
Added minor suggestions

File: modules/pubmatic/openwrap/cache/gocache/partner_config.go (hunk 1 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/cache/gocache/partner_config.go:100]: Done

File: modules/pubmatic/openwrap/database/mysql/adunit_config.go (hunk 2 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/database/mysql/adunit_config.go:25]: Added UT's

File: modules/pubmatic/openwrap/database/mysql/adunit_config.go [hunk 3]
```diff
@@ -20,29 +22,32 @@ func (db *mySqlDB) GetAdunitConfig(profileID, displayVersion int) (*adunitconfig
...
 		return nil, err
 	}
 
-	adunitConfig := &adunitconfig.AdUnitConfig{}
-	err = json.Unmarshal([]byte(adunitConfigJSON), &adunitConfig)
-	if err != nil {
-		return nil, adunitconfig.ErrAdUnitUnmarshal
-	}
+	if len(adunitConfigJSON) > 0 {
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/database/mysql/adunit_config.go:29]: Could you add UT to cover this change

**Review by pm-nilesh-chate — APPROVED** [review]
Approved with a minor comment

File: modules/pubmatic/openwrap/database/mysql/adunit_config.go (hunk 3 above)
Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/database/mysql/adunit_config.go:29]: Added UT, however UT  "No rows for given query" covers case of  len(adunitConfigJSON) = 0

**Review by pm-nilesh-chate — APPROVED** [review]

**Review by ashishshinde-pubm — APPROVED** [review]

File: modules/pubmatic/openwrap/cache/gocache/partner_config.go [hunk 4]
```diff
@@ -95,5 +96,8 @@ func (c *cache) getActivePartnerConfigAndPopulateWrapperMappings(pubID, profileI
 		c.metricEngine.RecordDBQueryFailure(queryType, strconv.Itoa(pubID), strconv.Itoa(profileID))
 		err = errorWrap(err, errAdunitConfig)
 	}
+	if errWrapperSlotMapping == nil && errAdunitConfig == nil {
```
Comment (pm-aadit-patil) [inline modules/pubmatic/openwrap/cache/gocache/partner_config.go:99]: There can be a case where adunit is not set for profile as it is optional, so MySQL returns _"sql: no rows in result set"_
as `populateCacheWithAdunitConfig()` internally calls `GetAdunitConfig()` which uses `conn.QueryRow()` 
QueryRow() returns error including `no rows in result` set as well

we can either change the implementation or of GetAdunitConfig()  or handle with ignoring 
`if errWrapperSlotMapping == nil && (errAdunitConfig == sql.ErrNoRows || errAdunitConfig == nil)`

Comment (pm-jaydeep-mohite) [reply modules/pubmatic/openwrap/cache/gocache/partner_config.go:99]: Error sql.ErrNoRows is handled at https://github.com/PubMatic-OpenWrap/prebid-server/blob/3295b496caa1b50e22780d8244ee58fc77316dc7/modules/pubmatic/openwrap/database/mysql/adunit_config.go#L25

**Review by pm-nilesh-chate — APPROVED** [review]

**Pull request merged**